src/ucp/core/ucp_ep.h

```diff
@@ -506,43 +511,37 @@ typedef struct {
 
 } ucp_ep_ext_control_t;
 
-
 /**
- * Endpoint extension for generic non fast-path data
+ * Endpoint extension
  */
-typedef struct {
-    void                          *user_data;    /* User data associated with ep */
-    ucs_list_link_t               ep_list;       /* List entry in worker's all eps list */
+struct ucp_ep_ext {
+    ucp_ep_h                 ep;            /* Back pointer to endpoint */
+    void                     *user_data;    /* User data associated with ep */
+    ucs_list_link_t          ep_list;       /* List entry in worker's all eps list */
     /* Endpoint match context and remote completion status are mutually exclusive,
      * since remote completions are counted only after the endpoint is already
      * matched to a remote peer.
      */
     union {
-        ucp_ep_match_elem_t       ep_match;      /* Matching with remote endpoints */
-        ucp_ep_flush_state_t      flush_state;   /* Remote completion status */
+        ucp_ep_match_elem_t  ep_match;      /* Matching with remote endpoints */
+        ucp_ep_flush_state_t flush_state;   /* Remote completion status */
     };
-    ucp_ep_ext_control_t          *control_ext;  /* Control data path extension */
+    ucp_ep_ext_control_t     control_ext;  /* Control data path extension */
```

Comment:
Maybe flatten the struct and get rid of `ucp_ep_ext_control_t`?

Reply:
yes, it would be better